### ISSUE TITLE
Using Asterisk or QualifiedAsterisk may produce unexpected columns when using it with generated code

### ISSUE THREAD

**Issue by stuppy** [jOOQ/jOOQ#8080]
### Expected behavior and actual behavior:
Adding new columns to a database table should not break existing code deployments with generated code.

When a new column is added, jooq indices no longer line up with the returned data -- possibly only when using `TABLE.asterisk()` -- and int/string parsing fails (at best; at worst columns are rearranged in the record).

Not really sure how to avoid breaking production unless we deploy new binary with code regenerated AND execute the alter query at exactly the same time.  Is there a way to turn on column-name lookups, rather than index?

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

Using `TABLE.asterisk()` in a query with generated code.
1. Add a new column to the database table.
2. Execute query again.

```
Caused by: java.sql.SQLException: Error while reading field: ... at JDBC index: 55
...
Caused by: java.sql.SQLException: Invalid value for getInt() - '<some string>'
```

### Versions:

- jOOQ: 3.11.5
- Java: 8
- Database (include vendor): MySQL 5.7
- OS: Cloud (App Engine)
- JDBC Driver (include name if inofficial driver): Cloud SQL

Thanks,
John

**Comment by stuppy** [comment]
Seems like multiple `.select(TABLE.fields()).select(OTHER_TABLE.fields())` is the workaround as it will only inline the known columns.  Downside is the multiple select calls, but they concat so all is well.

**Comment by lukaseder** [comment]
Thank you very much for your report. I'm assuming you inserted the new column somewhere in the middle? When you're using `Table.asterisk()`, you're explicitly requesting jOOQ to generate the `table.*` expression in generated SQL, which fetches more columns than jOOQ knew of, and in a different order.

This is how SQL works. If the client model of the table doesn't match the server model, there will be problems. Of course, there are workarounds:

- Explicitly listing all the columns as you did. This is always a better choice than working with asterisks. I would *always* avoid asterisks if possible, they're just trouble.
- Don't tell jOOQ what your columns are. You could of course use plain SQL to generate your asterisked expression, so jOOQ doesn't know what it is, so it will fetch the actual columns from JDBC's `ResultSetMetaData`

I'll think about whether we should in some cases work with `ResultSetMetaData` column names, rather than positional access, but there are two significant drawbacks of doing that:

- Column access performance is much worse for large result sets (this could be mitigated by creating a mapping only once per result set, but such a mapping isn't always applicable)
- Two columns may have ambiguous names in SQL, e.g. when joining or aliasing them. Name based column access doesn't work for all SQL statements

**Comment by lukaseder** [comment]
To this date, I couldn't think of any good reasons why the asterisk behaviour should be changed. When using the asterisk, the column order is explicitly defined by the server, and not by jOOQ. Supporting the asterisk was not a strict requirement in early jOOQ, and it was introduced precisely for that purpose, to allow relying on server side column ordering. Perhaps that's worth better documenting: https://github.com/jOOQ/jOOQ/issues/12620

There might be room for improving the `Table.fields()` usage, such that the expression can be used in a single `select()`, but that's just convenience. You could also wrap a table's contents in a nested `ROW` starting from jOOQ 3.15, see e.g. https://github.com/jOOQ/jOOQ/issues/11812.

I'll reject this issue here as "won't fix".
